### ISSUE TITLE
Ubershaders via N64 Rendering Pipeline Interpreter

### ISSUE THREAD

**Issue by retrobenny** [gonetz/GLideN64#1606]
Making this so it can be acknowledged if its reasonably doable.

If Dolphin can emulate shader storage of Gamecube and Wii via an Interpreter,then GLideN64 should be able to do this as well via an Interpreter of the N64's shader generation.

It would be excellent if we could have the same stutter-free experience in N64 emulation as we have in Dolphin for Gamecube and Wii games with a minimal performance decrease if 2 modes are also made for it of the hybrid and exclusive usage with hybrid having almost no performance impact.

The main caveat is the difficulty of creating such an Interpreter for shader generation.
There has to be enough info out there about N64 hardware etc. for creating a shader generation Interpreter.

**Comment by mudlord** [comment]
The Vulkan based renderer for the N64 core for RetroArch uses several compute shaders indeed for the combiner pipeline so no actual shader code is stitched together like in GlideN64 now.

**Comment by olivieryuyu** [comment]
Where do u have stutters? 

How would you really think emulating the rendering pipeline of the N64 would bring real benefit? How difficult to implement this matter for so, if existing, benefits? Dolphin ubershader is demanding on hardware, even on hybrid.

Honestly a good optimization of the code should be sufficient for our good old N64.

**Comment by retrobenny** [comment]
Cool,though that is LLE only and as a result extra slow compared to most any common HLE setup.
Trying that on Android (Shield TV) showed just how slow it still it when testing SM64 with it.

That is Vulkan though,I haven't seen anything about GLideN64 having a Vulkan branch,and Dolphin using Vulkan is weirdly slower than other backends instead of faster.
If it matches the conditions of Dolphin,a D3D GLideN64 would probably be faster,but that won't be considered as easily due to platform restriction unless someone dedicated full work on their own branch for it.

Zelda Ocarina seems to be the most likely game to have stutters for me,most notably on Android.
Make sure to start with fresh shaders or just have them disabled,also the stutters were last noticed on Android while Conker running on Android doesn't have any stuttering somehow since the last time I checked.
So this would be more openly an issue for Android than it is for Windows,at least I don't think I saw the same kind of stutters. (I never use shader storage due to the inherent risk of visual glitches if you mess with settings)
If you can,make a new save for Zelda and save the game when you first get control,then on next boot,if you go outside and turn the camera and also end up traveling to the Deku tree,there is a handful of easy stutters that usually happen,ignore the actual loading lag but there is a stutter before or after that moment.
I will have to boot up Zelda again later to check for stutters on the Android test build to see if they are still intact.

Also,N64 stutters are considerably shorter and less frequent than Gamecube/Wii emulation was.

**Comment by gonetz** [comment]
I do not understand, what is suggested here. What is shader generation Interpreter? Why we need it? Please explain it in few words, if possible.

**Comment by retrobenny** [comment]
This relates to the shader caching and the generation before they are cached.

**To explain the idea in brief**,make an accurate way of handling the N64 graphics pipeline in order to avoid the stutters of shader compilation by creating an Interpreter of the process,because interpreting them skips the step of compiling which avoids all chances of stutters happening.

More in depth;

The shader generation interpreter thought is about how shader compilation for rendering graphics and various effects can cause stuttering,compiling this can sometimes or oftenly take a long time and will cause those stutters if it takes too long for an uncached shader to compile from it.
The idea of using Interpreter on that part of GLideN64 instead is to execute the shaders natively instead of having to compile them before execution,this could technically be "slower" but would remove all stutters because it doesn't have to compile anything into the platform's language before execution.

Normally,Ubershaders are multiple powerful shader interpreters that directly execute the pipeline of the system they are made for instead of having to compile them into the platform's readable language associated with potential stutters from compilation time.
How Ubershaders work via hybrid mode is that hybrid still has the normal engine (compiler) intact for compiling the shaders on the side in a separate thread and reading them as stored/cached shaders on future appearances while the uber portion executes everything that is not already cached,thus avoiding stutters caused by fresh shader generation since it executes natively first before having the normal shader create a compiled version of it to store/cache it for direct use on the same normal shader the next time it gets generated.
Ubershaders with at least hybrid mode are multi-threaded due to how it works for less of an impact than exclusive mode but with no major consequence other than taking more space due to cache files like the standard shader compiler on GLideN64 does already if cache is enabled.

The Ubershader multi-threading makes it so that the normal shader can run on its own thread to compile and store/cache shaders that have already been executed then it would execute them as pre-compiled shaders instead of having the Uber portion execute its native generation again.

Dolphin with hybrid Ubershaders enabled runs with almost no performance impact,currently not too extreme of a slowdown when higher than native scale due to being optimized since released.
So the optimizations it received made most of the performance impacts go away.
Again,a version of this for N64 emulation would have a much lower impact due to not being as intense as those later consoles.

TL;DR Sorry for how crazy long and repetitive this may have gotten after my attempt of a brief explanation.

**Comment by retrobenny** [comment]
Here is the Dolphin article which explains the problem and part of how things work in general terms.

https://dolphin-emu.org/blog/2017/07/30/ubershaders/

Their ultimate fix in a nutshell.

![image](https://user-images.githubusercontent.com/16418716/31049009-11373918-a5f0-11e7-8712-213ba6b0a2ff.png)

**Comment by fzurita** [comment]
What he means is to stop stitching a bunch of shader code pieces and instead have one big shader that has all the logic inside. Since only one Fragment would have to be compiled, there would be no shader stutter.

**Comment by AmbientMalice** [comment]
Mobile devices are too slow for such a shader technique, and desktop devices arguably don't have significant enough stuttering issues to warrant it. The N64 and the GC/Wii are not comparable pieces of hardware.

**Comment by loganmc10** [comment]
Yeah personally I'd rather have the shader stutter than 1 giant shader. The stutter only happens the first time you play the game, then it's stored in binary format. An ubershader needs to have a bunch of if statements in it and it needs to be processed 100% of the time every time

**Comment by gonetz** [comment]
Ok, I got ihe idea,  thanks. 
Yes, it can be done for our project. It is not very hard and the uber will not very large. The question do we really need it is still open. Since I have 200+ open isues, most of which are more important than this one, I can only help with an advice.

**Comment by retrobenny** [comment]
Its more of a quality-of-life improvement.

@AmbientMalice Dolphin runs well enough on Shield TV when Ubershaders are enabled as explained on that blog article,plus I have tested it myself to see that performance was pretty smooth with them enabled.

If I remember correctly,the more features used,the heavier the stutter chances get.
I think after I set Anisotropy to the highest setting,the stutters became more noticeable,at least on Android when I looked at the title sequence of Banjo-Tooie as it kept on getting stutters every time a logo was about to appear on the screen.

Its not like the uber would completely replace the regular method,it would just be an option to toggle if you want to outright avoid stutters for a smoother experience and more accurate game pacing,like if you are about to perform an action,but it stutters and interrupts you.
I wonder,does unnoticeable (less than a frame) stuttering cause dropped inputs or input latency?

Zero stutters with GLideN64 on any N64 emulator is exactly like how the console would act,so it would match more closely with an accuracy increase for timing due to lack of stutters and also make it possible to avoid stutters with shader cache disabled so you could avoid caching issues related to changed settings or corrupted cache,hence why I leave shader cache disabled.

**Comment by LuismaSP89** [comment]
I agree with ubershaders + gliden64 since i can´t play goldeneye with all maxed settings  (1080p, 16x anisotropic, 8x border filter, 6xbrz, etc)  "Other games works well, but goldeneye with that settings is nearly unplayable, specially with the stutter when you select other weapon" (I got a powerful pc, that´s not the problem)

So, with ubershaders maybe can be solved many of the gliden64 framedrop issues (Just in my opinion)

**Comment by Eignar17** [comment]
@LuismaSP89 maybe can be and not fix your problem 

@retrobenny D3D in windows not fund on mac linux android or others sistems and vulkan is bat optimize where did you get that?

view real data
https://www.libretro.com/index.php/nintendo-64-vulkan-low-level-emulator-parallel-pre-alpha-release/
https://es.dolphin-emu.org/blog/2017/06/03/dolphin-progress-report-may-2017/

**Comment by oddMLan** [comment]
This can be closed, right? Shaded keys implemented

**Comment by gonetz** [comment]
Right. Shaders keys storage is better solution than ubershader.
